go/pools/smartconnpool/waitlist.go

```diff
@@ -0,0 +1,160 @@
+	age uint32
+}
+
+type waitlist[C Connection] struct {
+	nodes sync.Pool
+	mu    sync.Mutex
+	list  list.List[waiter[C]]
+}
+
+// waitForConn blocks until a connection with the given Setting is returned by another client,
+// or until the given context expires.
+// The returned connection may _not_ have the requested Setting. This function can
+// also return a `nil` connection even if our context has expired, if the pool has
+// forced an expiration of all waiters in the waitlist.
+func (wl *waitlist[C]) waitForConn(ctx context.Context, setting *Setting) (*Pooled[C], error) {
+	elem := wl.nodes.Get().(*list.Element[waiter[C]])
+	elem.Value = waiter[C]{setting: setting, conn: nil, ctx: ctx}
+
+	wl.mu.Lock()
+	// add ourselves as a waiter at the end of the waitlist
+	wl.list.PushBackValue(elem)
+	wl.mu.Unlock()
+
+	// block on our waiter's semaphore until somebody can hand over a connection to us
+	elem.Value.sema.wait()
+
+	// we're awake -- the conn in our waiter contains the connection that was handed
+	// over to us, or nothing if we've been waken up forcefully. save the conn before
+	// we return our waiter to the pool of waiters for reuse.
+	conn := elem.Value.conn
+	wl.nodes.Put(elem)
+
+	return conn, ctx.Err()
+}
+
+// expire removes and wakes any expired waiter in the waitlist.
+// if force is true, it'll wake and remove all the waiters.
+func (wl *waitlist[C]) expire(force bool) (waiting int) {
+	if wl.list.Len() == 0 {
+		return 0
+	}
+
+	var expired []*list.Element[waiter[C]]
+
+	wl.mu.Lock()
+	// iterate the waitlist looking for waiters with an expired Context,
+	// or remove everything if force is true
+	for e := wl.list.Front(); e != nil; e = e.Next() {
+		if force || e.Value.ctx.Err() != nil {
+			wl.list.Remove(e)
+			expired = append(expired, e)
+			continue
+		}
+		waiting++
+	}
+	wl.mu.Unlock()
+
+	// once all the expired waiters have been removed from the waitlist, wake them up one by one
+	for _, e := range expired {
+		e.Value.sema.notify(false)
+	}
+	return waiting
+}
+
+// tryReturnConn tries handing over a connection to one of the waiters in the pool.
+func (wl *waitlist[D]) tryReturnConn(conn *Pooled[D]) bool {
+	// fast path: if there's nobody waiting there's nothing to do
+	if wl.list.Len() == 0 {
+		return false
+	}
+	// split the slow path into a separate function to enable inlining
+	return wl.tryReturnConnSlow(conn)
+}
+
+func (wl *waitlist[D]) tryReturnConnSlow(conn *Pooled[D]) bool {
+	const maxAge = 8
+	var (
+		target      *list.Element[waiter[D]]
+		connSetting = conn.Conn.Setting()
+	)
+
+	wl.mu.Lock()
+	target = wl.list.Front()
+	// iterate through the waitlist looking for either waiters that have been
+	// here too long, or a waiter that is looking exactly for the same Setting
+	// as the one we have in our connection.
+	for e := target; e != nil; e = e.Next() {
+		if e.Value.age > maxAge || e.Value.setting == connSetting {
+			target = e
+			break
+		}
+		e.Value.age++
+	}
```

Comment:
here were are only ageing the top waiters where as other waiters could also be ageing which is ignored if any waiter in the middle matches the setting.

Reply:
not sure if this should be solved, but may be `maxAge` can be reduced.

Reply:
Yeah good catch, I added a comment: skipping the aging of the back of the list is intentional because we only want to "age" a client when it's being skipped over by other requests. The goal was not "total age of the waiter" but instead "times it's been skipped over". The `maxAge` of 8 was set by testing different values. It seems to work well in practice, because smaller values cause connections-with-settings to starve slightly, and aging _all the waiters_ every time leads to unfairness.